realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp

```diff
@@ -1828,3 +1829,22 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeGetFinalizerPtr(
     TR_ENTER()
     return reinterpret_cast<jlong>(&finalize_table_query);
 }
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_TableQuery_nativePredicate(JNIEnv *env, jobject, jlong nativeQueryPtr, jstring j_filter)
+{
+    TR_ENTER_PTR(nativeQueryPtr);
+    try {
+        Query* query = reinterpret_cast<Query *>(nativeQueryPtr);
+        JStringAccessor filter(env, j_filter); // throws
+        parser::ParserResult parser_result = realm::parser::parse(filter);
+
+        query_builder::NoArguments no_args;
+        query_builder::apply_predicate(*query, parser_result.predicate, no_args);
+
+        // TODO What about ordering. What does this mean?
+        // DescriptorOrdering ordering;
+        // realm::parser::query_builder::apply_ordering(ordering, table, parser_result.ordering);
```

Comment:
The ordering handles sort/distinct (and soon limit) parts of the query. I think to handle that here, you'd want to save the ordering produced by `apply_ordering` and then apply it when the user calls `find_all()` or `count()` etc. I'm guessing there is already a ordering state somewhere since it is supported through Java's typed query API, so you could append the sort/distinct descriptors of this result with the existing state (the API's for this aren't super friendly right now but it is possible).

Reply:
Thanks 👍 Yes, we are storing a Descriptor object in Java right now. It will be a bit fiddly trying having to send this one back from this and then decide which to use if people later override it, but it isn't impossible. Thanks for the clarification 😄

Reply:
It should be possible to merge orderings, so that all of them are applied. This should let users combine sort/distinct across the typed and untyped API correctly and both the sort/distinct parts be applied in the requested order.

```
realm.where(Person.class)
  .equalTo("name", "Jane")
  .filtered("children.@count > 2 SORT(name ASC) DISTINCT(name)")
  .lessThan("age", 35)
  .sort("age", ASCENDING)
  .distinct("age")
  .findAll();
```

It would be a bit tricky to do this currently, but we could easily add a `DescriptorOrdering::appendOther(DescriptorOrdering other)` method in core.

Reply:
Right now we are storing a separate `sortDescriptor` and `distinctDescriptor` in Java, keeping the latest one defined and when you call `findAll()` it merges them using `append_sort` and `append_distinct` on a new Descriptor object.

I assume the same approach would work for the `ParseResult`? Although it doesn't seem possible to pull out the sort/distinct parts of the Descriptor individually?

If that isn't the case, having a method that merges them seems required.  Then we could also simplify the logic in the Java query class and only have one Descriptor.